lib/collection/src/collection/shard_transfer.rs

```diff
@@ -257,33 +256,37 @@ impl Collection {
             .stop_task(&transfer_key)
             .await;
 
-        let replica_set =
-            if let Some(replica_set) = shard_holder_guard.get_shard(&transfer_key.shard_id) {
-                replica_set
-            } else {
-                return Err(CollectionError::bad_request(format!(
-                    "Shard {} doesn't exist",
-                    transfer_key.shard_id
-                )));
-            };
+        let mut shard_holder_guard = None;
+
+        let shard_holder = match shard_holder {
+            Some(shard_holder) => shard_holder,
+            None => shard_holder_guard.insert(self.shards_holder.read().await),
+        };
 
-        let transfer = shard_holder_guard.get_transfer(&transfer_key);
+        let Some(replica_set) = shard_holder.get_shard(&transfer_key.shard_id) else {
+            return Err(CollectionError::bad_request(format!(
+                "Shard {} doesn't exist",
+                transfer_key.shard_id
+            )));
+        };
+
+        let Some(transfer) = shard_holder.get_transfer(&transfer_key) else {
+            return Ok(());
+        };
 
-        if transfer.map(|x| x.sync).unwrap_or(false) {
-            replica_set.set_replica_state(&transfer_key.to, ReplicaState::Dead)?;
-        } else {
-            replica_set.remove_peer(transfer_key.to).await?;
+        if replica_set.peer_state(&transfer.to).is_some() {
+            if transfer.sync {
+                replica_set.set_replica_state(&transfer.to, ReplicaState::Dead)?;
+            } else {
+                replica_set.remove_peer(transfer.to).await?;
+            }
         }
```

Comment:
We only change replica state/remove peer if it *exists* now.

Before we would `set_replica_state(Dead)` even if peer is already removed, which would *re-create* the peer. And then the peer would request shard transfer, because it's `Dead`, so it was non-trivial to cancel `sync = true` shard transfer.

We also only call `remove_peer` if shard transfer *exists*.

Before it was possible to accidentally delete the wrong replica, if multiple `Abort` messages were sent for a *non* `sync = true` transfer, because `.unwrap_or(false)` would trigger `remove_peer` branch if `transfer = None`.